fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/FileProvider.java

```diff
@@ -79,43 +83,104 @@ public long getSize(String workItem) throws FHIRException {
+                            // Log and skip the invalid FHIR resource.
+                            parseFailures++;
+                            logger.log(Level.INFO, "readResources: Failed to parse line " + (numOfLinesToSkip + chunkRead) + " of [" + source + "].", e);
+                        }
+
+                        // With small resources the 50 at a time gets really expensive.
+                        // so we have a READ_BLOCK_OPT to get at least a decent segment to
+                        // insert into the db.
+                        if (chunkRead < maxRead || reader.getLength() < READ_BLOCK_OPT) {
+                            resourceStr = reader.readLine();
+                        } else {
+                            resourceStr = null;
+                            continueRead = false;
+                        }
                     }
-                    resourceStr = br.readLine();
-                }
-            } finally {
-                if (br != null) {
-                    br.close();
+                    // The Channel enables us to skip the future seek.
+                    this.transientUserData.setCurrentBytes(this.transientUserData.getCurrentBytes() + reader.getLength());
                 }
             }
         } catch (Exception e) {
             throw new FHIRException("Unable to read from Local File", e);
         }
     }
 
+    /**
+     * This specialized class enables a BufferedReader.readLine like implementation
+     * without the buffering as there is a local file with direct read and seek.
+     */
+    private static class CountInputStreamReader extends InputStreamReader {
+        // Implies CR
+        private static final int LF = '\n';
+        private static final long MAX_LENGTH_PER_LINE = 2147483648l;
+
+        private long length = 0;
+
+        public CountInputStreamReader(InputStream in) {
+            super(in);
+        }
+
+        /**
+         * Read the line
+         * @return
+         * @throws IOException
+         */
+        public String readLine() throws IOException {
+            int r = this.read();
+            if (r == -1) {
+                // End-of-stream
+                return null;
+            }
+            boolean read = true;
+            StringBuilder builder = new StringBuilder();
+            int lineLength = 0;
+
+            // Protect against attacks with a max line length (and max size we support in the db).
+            while (read && lineLength < MAX_LENGTH_PER_LINE) {
+                if (r == -1) {
+                    read = false;
+                } else if (r == LF) {
+                    // \n  case
+                    length++;
+                    read = false;
+                } else {
+                    builder.append((char) r);
+                    length++;
+                    // We only need to account for line length here.
+                    lineLength++;
+                    r = this.read();
```

Comment:
Could be very slow unless the underlying InputStream is buffered.

Reply:
This code is based on the BufferedReader. It follows a similar pattern.